api/src/main/java/org/openmrs/api/impl/ProgramWorkflowServiceImpl.java

```diff
@@ -261,7 +261,12 @@ public PatientProgram savePatientProgram(PatientProgram patientProgram) throws A
 				}
 			}
 		}
-		
+		// Makes sure that the end dates of most recent states in each workflow
+		// and the program end date are consistent
+		for (PatientState state : patientProgram.getMostRecentStateInEachWorkflow()) {
```

Comment:
I thought this should only be happening when completing a program, am not seeing the condition that ensures that.

Reply:
@wluyima this confused me for a second as well, but it's actually more or less as the ticket specifies, because it's handling these two cases:

* When completing a program, the end date of the most recent state in each workflow should be set to the end date of the program
* When uncompleting a program (setting end date to null) the end date of the most recent state in each workflow should be set to null

However, I'm not sure anymore if this is the correct behavior.

Reply:
This method is called to save or update a program so not all the time the caller is completing or uncompleting a program

Reply:
Fair enough... will add some more comments on the ticket.

Reply:
Based on our discussion on this ticket can we change this loop so that it only executes if patientProgram.getDateCompleted() == null, and then it only sets the endDate for state if state.setEndDate() == null?

I think that's the only change we will need, though we'd want to add some tests for these cases.  Sorry for the additional changes @shifona1, let me know if you have any questions.

Reply:
Thanks @mogoodrich @wluyima ,
I'll update the changes soon. 
So, just to confirm, It should be ensured that the loop executes only when the program is completed or uncompleted and only the recent states which have their enddate as null (that is they are not suspended) should be set. 

The completion of the program can be detected if the enddate is set i.e `patientProgram.getDateCompleted() != null`.
Also, from what i have understood currently, we are not handling the case when a program is uncompleted, so we will also need to remove the test cases related to the same, since, they will fail otherwise. 
So, there are roughly three major changes: one in this file, second to add tests for these cases and third to remove the tests for uncompletion.

Reply:
Yes, that sounds correct!  And to your other questions, it is fine to have the tests in ProgramWorkflowServiceTests... this does seem like the correct place for them.